src/SubscriberBase.h

```diff
@@ -18,34 +30,115 @@ class SubscriberBaseT : public Subscriber<T>,
   virtual void onCompleteImpl() = 0;
   virtual void onErrorImpl(folly::exception_wrapper ex) = 0;
 
+  // used to be able to cancel subscription immediatelly, making sure we dont
+  // deliver any other signals after that
+  // also to break the reference cycle involving storing subscription pointer
+  // for the users of the SubscriberBase
+  class SubscriptionShim : public Subscription {
```

Comment:
Why doesn't this class wrap a `Subscription` instead of a subscriber? Seems like it makes more sense to have a `Subscription` which accepts an `EventBase`, and ensures that cancellation makes other kinds of messages invalid.